File: latest_ran_output/gpt_oss_120b/auth_004_naive.py
import os
import datetime
from functools import wraps

import bcrypt
import jwt
from flask import Flask, request, jsonify, make_response, g

app = Flask(__name__)
app.config["SECRET_KEY"] = os.getenv("FLASK_SECRET", "super-secret-key")
# In a real app use a secure random key and keep it out of source control.

# ----------------------------------------------------------------------
# Mock user store (replace with DB in production)
# Passwords are stored as bcrypt hashes.
# ----------------------------------------------------------------------
_USERS = {
    "alice": bcrypt.hashpw(b"password123", bcrypt.gensalt()).decode(),
    "bob":   bcrypt.hashpw(b"mysecurepwd", bcrypt.gensalt()).decode(),
}

def _verify_password(stored_hash: str, password: str) -> bool:
    return bcrypt.checkpw(password.encode(), stored_hash.encode())

# ----------------------------------------------------------------------
# JWT helpers
# ----------------------------------------------------------------------
def _generate_token(username: str, remember_me: bool) -> str:
    """Create a JWT token. 30‑day expiry if remember_me else 1‑hour."""
    now = datetime.datetime.utcnow()
    exp_delta = datetime.timedelta(days=30) if remember_me else datetime.timedelta(hours=1)
    payload = {
        "sub": username,
        "iat": now,
        "exp": now + exp_delta,
    }
    token = jwt.encode(payload, app.config["SECRET_KEY"], algorithm="HS256")
    # PyJWT returns str in >=2.0, bytes in <2.0 – ensure string.
    return token if isinstance(token, str) else token.decode()

def _decode_token(token: str):
    try:
        payload = jwt.decode(token, app.config["SECRET_KEY"], algorithms=["HS256"])
        return payload
    except jwt.ExpiredSignatureError:
        raise PermissionError("Token expired")
    except jwt.InvalidTokenError:
        raise PermissionError("Invalid token")

# ----------------------------------------------------------------------
# Request hooks & decorators
# ----------------------------------------------------------------------
@app.before_request
def load_current_user():
    """Load user from JWT cookie into `g.current_user` if present."""
    token = request.cookies.get("access_token")
    if not token:
        g.current_user = None
        return
    try:
        payload = _decode_token(token)
        g.current_user = payload["sub"]
    except PermissionError:
        g.current_user = None

def login_required(f):
    @wraps(f)
    def decorated(*args, **kwargs):
        if not getattr(g, "current_user", None):
            return jsonify({"error": "Authentication required"}), 401
        return f(*args, **kwargs)
    return decorated

# ----------------------------------------------------------------------
# Routes
# ----------------------------------------------------------------------
@app.route("/login", methods=["POST"])
def login():
    """
    Expected JSON:
    {
        "username": "...",
        "password": "...",
        "remember_me": true|false   (optional, defaults to false)
    }
    """
    data = request.get_json(silent=True) or {}
    username = data.get("username")
    password = data.get("password")
    remember_me = bool(data.get("remember_me", False))

    if not username or not password:
        return jsonify({"error": "Username and password required"}), 400

    stored_hash = _USERS.get(username)
    if not stored_hash or not _verify_password(stored_hash, password):
        return jsonify({"error": "Invalid credentials"}), 401

    token = _generate_token(username, remember_me)

    resp = make_response(jsonify({"message": "Logged in"}))
    # HttpOnly + Secure (if using HTTPS) cookie
    resp.set_cookie(
        "access_token",
        token,
        httponly=True,
        secure=False,          # set True when serving over HTTPS
        samesite="Lax",
        max_age=30 * 24 * 3600 if remember_me else None,
    )
    return resp

@app.route("/logout", methods=["POST"])
def logout():
    """Clear the JWT cookie."""
    resp = make_response(jsonify({"message": "Logged out"}))
    resp.delete_cookie("access_token")
    return resp

@app.route("/profile")
@login_required
def profile():
    """Example protected endpoint."""
    return jsonify({
        "user": g.current_user,
        "msg": f"Welcome back, {g.current_user}!"
    })

# ----------------------------------------------------------------------
# Run (for development only)
# ----------------------------------------------------------------------
if __name__ == "__main__":
    app.run(debug=True)

File: latest_ran_output/gpt_oss_120b/auth_004_naive.js
// server.js
const express = require('express');
const bcrypt = require('bcrypt');
const jwt = require('jsonwebtoken');
const cookieParser = require('cookie-parser');
const bodyParser = require('body-parser');

// ---- Config ----------------------------------------------------
const JWT_SECRET = 'your-very-secure-secret'; // replace with env var in prod
const JWT_ISSUER = 'myapp';
const PORT = process.env.PORT || 3000;

// ---- Mock DB (replace with real DB) ----------------------------
const users = [
  // password is bcrypt hash of "password123"
  {
    id: 1,
    email: 'alice@example.com',
    passwordHash:
      '$2b$10$KIX/6G9ZcXx5UeVv3FQhOeY8pJkWzR7gM0EwBfHjL4nGd1yN2aK5u', // bcrypt hash
    name: 'Alice',
  },
];

// ---- Helpers ---------------------------------------------------
function findUserByEmail(email) {
  return users.find((u) => u.email === email);
}

function generateToken(payload, rememberMe = false) {
  const expiresIn = rememberMe ? '30d' : '1h';
  return jwt.sign(payload, JWT_SECRET, {
    expiresIn,
    issuer: JWT_ISSUER,
  });
}

// ---- Middleware ------------------------------------------------
function authMiddleware(req, res, next) {
  const token = req.cookies['auth_token'];
  if (!token) return res.status(401).json({ error: 'Unauthenticated' });

  try {
    const decoded = jwt.verify(token, JWT_SECRET, { issuer: JWT_ISSUER });
    req.user = decoded; // attach payload (e.g., id, email)
    next();
  } catch (err) {
    return res.status(401).json({ error: 'Invalid or expired token' });
  }
}

// ---- App Setup -------------------------------------------------
const app = express();
app.use(bodyParser.json());
app.use(cookieParser());

// ---- Routes ----------------------------------------------------
/**
 * POST /login
 * Body: { email, password, rememberMe? }
 */
app.post('/login', async (req, res) => {
  const { email, password, rememberMe } = req.body;
  if (!email || !password) {
    return res.status(400).json({ error: 'Email and password required' });
  }

  const user = findUserByEmail(email);
  if (!user) {
    return res.status(401).json({ error: 'Invalid credentials' });
  }

  const match = await bcrypt.compare(password, user.passwordHash);
  if (!match) {
    return res.status(401).json({ error: 'Invalid credentials' });
  }

  // Create JWT
  const token = generateToken(
    { id: user.id, email: user.email, name: user.name },
    rememberMe === true
  );

  // Set cookie (HttpOnly, SameSite)
  const cookieOptions = {
    httpOnly: true,
    sameSite: 'strict',
    secure: process.env.NODE_ENV === 'production', // true in prod
    maxAge: rememberMe ? 30 * 24 * 60 * 60 * 1000 : 1 * 60 * 60 * 1000, // ms
    path: '/',
  };
  res.cookie('auth_token', token, cookieOptions);

  return res.json({ message: 'Logged in successfully' });
});

/**
 * POST /logout
 */
app.post('/logout', (req, res) => {
  res.clearCookie('auth_token', { path: '/' });
  return res.json({ message: 'Logged out' });
});

/**
 * GET /profile - protected route example
 */
app.get('/profile', authMiddleware, (req, res) => {
  // In real app fetch fresh user data from DB using req.user.id
  const user = findUserByEmail(req.user.email);
  if (!user) return res.status(404).json({ error: 'User not found' });

  const { passwordHash, ...safeInfo } = user;
  return res.json({ profile: safeInfo });
});

// ---- Server ----------------------------------------------------
app.listen(PORT, () => {
  console.log(`Server listening on http://localhost:${PORT}`);
});